wasm/src/main/scala/ir2wasm/WasmExpressionBuilder.scala

```diff
@@ -1514,4 +1550,93 @@ private class WasmExpressionBuilder private (
     instrs += CALL(FuncIdx(WasmFunctionName.jsLinkingInfo))
     IRTypes.AnyType
   }
+
+  // ===============================================================================
+  // array
+  // ===============================================================================
+  private def genArrayLength(t: IRTrees.ArrayLength): IRTypes.Type = {
+    genTreeAuto(t.array)
+    instrs += ARRAY_LEN
+    IRTypes.IntType
+  }
+
+  private def genNewArray(t: IRTrees.NewArray): IRTypes.Type = {
+    if (t.lengths.isEmpty || t.lengths.sizeIs > t.typeRef.dimensions)
+      throw new AssertionError(
+        s"invalid lengths ${t.lengths} for array type ${t.typeRef.displayName}"
+      )
+
+    def genWithDimensions(typeRef: IRTypes.ArrayTypeRef, lengths: List[IRTrees.Tree]): Unit = {
+      val elemTy = extractArrayElemType(t.typeRef)
+      val wasmElemTy = TypeTransformer.transformType(elemTy)(ctx)
+      val arrTyName = WasmTypeName.WasmArrayTypeName(IRTypes.ArrayType(typeRef))
+      ctx.addArrayType(
+        WasmArrayType(
+          arrTyName,
+          WasmStructField(Names.WasmFieldName.arrayField, wasmElemTy, isMutable = true)
+        )
+      )
+      if (lengths.tail.isEmpty) {
+        val zero = Defaults.defaultValue(wasmElemTy)
+        instrs += zero
+      } else {
+        genWithDimensions(
+          IRTypes.ArrayTypeRef(typeRef.base, typeRef.dimensions - 1),
+          lengths.tail
+        )
+      }
+      genTree(lengths.head, IRTypes.IntType)
+      instrs += ARRAY_NEW(TypeIdx(arrTyName))
+    }
+    genWithDimensions(t.typeRef, t.lengths)
+    t.tpe
+  }
+
+  /** For getting element from an array, array.set should be generated by transformation of
+    * `Assign(ArraySelect(...), ...)`
+    */
+  private def genArraySelect(t: IRTrees.ArraySelect): IRTypes.Type = {
+    val arrayType = t.array.tpe match {
+      case t: IRTypes.ArrayType => t
+      case _ =>
+        throw new IllegalArgumentException(
+          s"ArraySelect.array must be an array type, but has type ${t.array.tpe}"
+        )
+    }
+    genTreeAuto(t.array)
+    genTree(t.index, IRTypes.IntType)
+    instrs += ARRAY_GET(TypeIdx(WasmTypeName.WasmArrayTypeName(arrayType)))
+
+    val typeRef = arrayType.arrayTypeRef
+    if (typeRef.dimensions > 1) IRTypes.ArrayType(typeRef.copy(dimensions = typeRef.dimensions - 1))
+    else
+      typeRef.base match {
+        case IRTypes.ClassRef(className) => ClassType(className)
+        case IRTypes.PrimRef(tpe)        => tpe
+      }
+  }
+
+  private def genArrayValue(t: IRTrees.ArrayValue): IRTypes.Type = {
+    val irElemTy = extractArrayElemType(t.typeRef)
+    val wasmElemTy = TypeTransformer.transformType(irElemTy)(ctx)
+    val arrTyName = Names.WasmTypeName.WasmArrayTypeName(t.tpe)
+    ctx.addArrayType(
+      WasmArrayType(
+        arrTyName,
+        WasmStructField(Names.WasmFieldName.arrayField, wasmElemTy, isMutable = true)
+      )
+    )
```

Comment:
These 9 lines are copy-pasted between here (`genArrayValue`) and `genNewArray`. Consider factoring them directly inside a method of `ctx`, for example
```scala
def getArrayTypeName(typeRef: IRTypes.ArrayTypeRef): WasmArrayTypeName = ...
```